File: day4/part-one-two.py
#
# Learned about strip(), very important! Now I'm confused as to whether I should
# use learned or learnt, dilemmas. Learnt about startswith(), and str() for
# printing ints.
#
# This is a pretty naive solution, might come back to it later
#

import sys
import hashlib

def get_key_part(secret_key_file):
    secret_key = secret_key_file.read().strip()
    key_part = 0
    digest = ''
    while not digest.startswith('000000'):
        key_part += 1
        digest = hashlib.md5(secret_key + str(key_part)).hexdigest()

    return key_part

if len(sys.argv) > 1:
    file_name = sys.argv[1]
else:
    file_name = 'input.txt'
secret_key_file = open(file_name, 'r')
print get_key_part(secret_key_file)
secret_key_file.close()

File: day4/part-one-two.js
/*
 * Forgot to add trim(), very important! Learnt about startsWith() and the
 * crypto node module.
 */

'use strict';
const fs = require('fs');
const crypto = require('crypto');

let fileName = process.argv[2] || 'input.txt';
fs.readFile(fileName, (err, data) => {
  console.log(getKeyPart(String(data).trim()));
});

function getKeyPart(secretKey) {
  let keyPart = 0;
  let digest = '';
  while (!digest.startsWith('000000')) {
    keyPart++;
    let md5sum = crypto.createHash('md5');
    md5sum.update(secretKey + keyPart);
    digest = md5sum.digest('hex');
  }

  return keyPart;
}
